### PR TITLE
Forward declared structs now generate opaque enums

### LINKED ISSUE #62: Detect forward declarations and generate opaque enums for them.
We should be able to detect forward declarations in bindgen, and generate opaque enums if appropriate, instead of structs with an address byte.

We should check `clang_getCursorDefinition()` and `clang_isCursorDefinition` to check whether what we're scanning, mark the generated type accordingly, and do the proper codegen steps.

### PR DESCRIPTION
**Pull request by cynicaldevil**
@emilio : I checked the test outputs again, and it seems that these changes are affecting struct *definitions* as well. Hence, I have not committed the test changes yet.
Fixes #62

### CODE REVIEW

File: libbindgen/src/ir/comp.rs [hunk 1]
```diff
@@ -236,6 +236,8 @@ pub struct CompInfo {
     /// Used to detect if we've run in a has_destructor cycle while cycling
     /// around the template arguments.
     detect_has_destructor_cycle: Cell<bool>,
+
+    is_declaration: bool,
```
Comment (emilio) [inline libbindgen/src/ir/comp.rs]: let's rename this and the method to `is_forward_declaration`, and also add a doc comment here describing what this indicates.

File: libbindgen/src/ir/comp.rs [hunk 2]
```diff
@@ -506,6 +509,7 @@ impl CompInfo {
         debug!("CompInfo::from_ty({:?}, {:?})", kind, cursor);
 
         let mut ci = CompInfo::new(kind);
+        ci.is_declaration = cursor.definition().unwrap().is_declaration();
```
Comment (emilio) [inline libbindgen/src/ir/comp.rs]: This is maybe not enough? The definition is _also_ a declaration, we need to search for a declaration that has no definition.

Note that clang has `CXCursor_NoDeclFound`, so we can probably do something like:

```rust
ci.is_forward_declaration = cursor.definition().map_or(true, |d| d.kind() == CXCursor_NoDeclFound);
```

That _might_ be enough.

Comment (cynicaldevil) [inline libbindgen/src/ir/comp.rs]: Why should the default value be true here?

Comment (cynicaldevil) [inline libbindgen/src/ir/comp.rs]: what if we modify the `from_ty` function to accept a cursor as well? We could simple use `clang_isDeclaration()` then.

**Comment by cynicaldevil** [comment]
Is there any way I can print out the cursor kind and run transformations on a single header file? This might give us some more information.

**Comment by emilio** [comment]
You can run bindgen on a header file with the --emit-clang-ast option,
and also with RUST_LOG=libbindgen to see pretty much everything you
could need.

**Comment by cynicaldevil** [comment]
`cursor.defintion()` *always* returns a cursor of type `StructDecl`, it doesn't matter whether the cursor points to a struct definition or a declaration. 

One solution, like I suggested earlier, is to modify the `CompInfo::from_ty` method to accept a cursor as well, and use that cursor to determine whether it points to a declaration or not.

**Comment by cynicaldevil** [comment]
@emilio I got two suggestions from twitter, which work somewhat:

1. Use `cursor.clang_isCursorDefinition()`, but this doesn't work for templated structs. No false positives here.
2. Parse the cursor as a token using `clang_tokenize()` and check if the name ends with a semicolon.

Option 2 is a hack, but it could be useful. Should I try using a combination of these two see if it works?

**Comment by emilio** [comment]
clang_isCursorDefinition sounds great, I'd rather patch libclang than tokenize, so lets use that :)

**Comment by emilio** [comment]
BTW, clang_isCursorDefinition was mentioned in the issue, I assumed you'd tried it :)

**Comment by cynicaldevil** [comment]
I'd tried it initally for the cursor we obtain from `ty.declaration()` in `from_ty`, but forgot to test it for the one in `from_clang_ty`.

File: libbindgen/tests/expectations/tests/namespace.rs [hunk 3]
```diff
@@ -66,6 +66,7 @@ pub mod root {
             #[link_name = "_ZN1w3fooEv"]
             pub fn foo() -> root::C<::std::os::raw::c_int>;
         }
+        pub enum C { }
```
Comment (cynicaldevil) [inline libbindgen/tests/expectations/tests/namespace.rs]: I don't know why this is appearing here, even though the struct has already been declared a few lines above.

**Comment by cynicaldevil** [comment]
@emilio I've updated the PR, but like I said, there's still a few false negatives that I can't get around. For example, `template < typename > struct char_traits;` does not resolve into an enum, because the corresponding CursorKind is `ClassTemplate`.

File: libbindgen/tests/expectations/tests/anon_union.rs [hunk 4]
```diff
@@ -42,18 +42,8 @@ pub const TErrorResult_UnionState_HasException: TErrorResult_UnionState =
...
-    pub _address: u8,
-    pub _phantom_0: ::std::marker::PhantomData<T>,
-}
-#[repr(C)]
-#[derive(Debug, Copy, Clone)]
-pub struct TErrorResult_DOMExceptionInfo<T> {
-    pub _address: u8,
-    pub _phantom_0: ::std::marker::PhantomData<T>,
-}
+pub enum TErrorResult_Message { }
```
Comment (emilio) [inline libbindgen/tests/expectations/tests/anon_union.rs]: This is wrong, you still need to make it take a template parameter, there is no way this is compiling otherwise.

File: libbindgen/src/codegen/mod.rs [hunk 5]
```diff
@@ -723,6 +723,17 @@ impl CodeGenerator for CompInfo {
             return;
         }
 
+        // generate an opaque enum if struct is a forward declaration
+        if self.kind() == CompKind::Struct && self.is_forward_declaration() {
```
Comment (emilio) [inline libbindgen/src/codegen/mod.rs]: Given that, let's skip also structs with template parameters.

Why only check for `Struct`? instead of both struct and union?

File: libbindgen/src/clang.rs [hunk 6]
```diff
@@ -190,6 +190,10 @@ impl Cursor {
         unsafe { clang_getCursorKind(self.x) }
     }
 
+    pub fn is_definition(&self) -> bool {
+        unsafe { clang_isCursorDefinition(self.x) == 0 }
```
Comment (emilio) [inline libbindgen/src/clang.rs]: This is returning true if the cursor is _not_ a definition, which causes a lot of confusion. Can you change this check to `!= 0`, and invert the conditional it in other places?

Also, this is going to need a line of documentation or this won't pass CI.

File: libbindgen/src/ir/comp.rs [hunk 7]
```diff
@@ -506,6 +512,8 @@ impl CompInfo {
         debug!("CompInfo::from_ty({:?}, {:?})", kind, cursor);
 
         let mut ci = CompInfo::new(kind);
+        ci.is_forward_declaration = location.map_or(true, |cur|
+                                        cur.is_definition() && cur.kind() == CXCursor_StructDecl);
```
Comment (emilio) [inline libbindgen/src/ir/comp.rs]: Why only if the cursor is `CXCursor_StructDecl`?

Also, I think this should work ok if you just do:

```rust
ci.is_forward_declaration = !cursor.is_definition();
```

That should make it work for structs with template parameters too. Does that work?

Comment (emilio) [inline libbindgen/src/ir/comp.rs]: Alternatively, worth trying `!cursor.definition().is_valid()`.

**Comment by cynicaldevil** [comment]
What happened here??!! :P 

image

**Comment by fitzgen** [comment]
> What happened here??!! :P

:) Everything that was in `./libbindgen/*` should now be in `./*` for the most part :P

**Comment by briansmith** [comment]
See also https://github.com/rust-lang/rfcs/pull/1861 and the referenced internals discussion about why using empty enums for this is (probably) wrong.

**Comment by fitzgen** [comment]
> See also rust-lang/rfcs#1861 and the referenced internals discussion about why using empty enums for this is (probably) wrong.

I'm not 100% sure that argument applies since we would be manipulating *pointers* to such enums, not the enum directly.

**Comment by cynicaldevil** [comment]
@briansmith We decided to go ahead with this and if needed, change it later :)

File: src/ir/comp.rs [hunk 8]
```diff
@@ -481,6 +486,9 @@ impl CompInfo {
         debug!("CompInfo::from_ty({:?}, {:?})", kind, cursor);
 
         let mut ci = CompInfo::new(kind);
+        ci.is_forward_declaration = location.map_or(true, |cur|
+                                        (cur.kind() == CXCursor_StructDecl || cur.kind() == CXCursor_UnionDecl) &&
```
Comment (cynicaldevil) [inline src/ir/comp.rs]: @emilio It might be a good idea to add a test for unions as well, since none of the existing tests are testing it.

**Comment by briansmith** [comment]
> @briansmith We decided to go ahead with this and if needed, change it later :)

If rust-bindgen starts using the empty enum pattern then it will create a lot of inertia preventing any deprecation of that pattern, which would be bad. Why not use the similar thing that the Rust lang team recommends instead, according to https://github.com/rust-lang/rfcs/pull/1861#issuecomment-274613536:
```
#[repr(C)]
pub struct Opaque([u8; 0]);
```

**Comment by emilio** [comment]
That sounds fine to me Brian. @cynicaldevil, it shouldn't be hard to change, instead of EnumBuilder you can use the tuple_struct method from aster, or the quote_item! macro, at your choice.

There are examples of both in that same file.

Thanks for commenting here @briansmith, I was the one initially proposing opaque enums but I wasn't aware of the issues.

Whether those issues are more or less likely to happen in practice for bindgen doesn't seem too relevant if there's a better and equally simple to implement alternative.

Thanks again!

**Comment by cynicaldevil** [comment]
@emilio pushed new changes.

**Comment by emilio** [comment]
I'm surprised this only hits once, may you add a few more tests for this? The common patters of forward-declaring a struct or union, then defining functions taking a pointer to that should be enough.

**Comment by cynicaldevil** [comment]
How about this:
```c++
struct Foo;

struct Bar {
    Foo *f;
};

void baz_struct(Foo* f);

union Union;

void baz_union(Union* u);

// conversion to tuple struct does not work on classes
class Quux;
```

**Comment by emilio** [comment]
That looks quite fine to me. We should be able to make it work on classes without template parameters though. But we can leave that as a followup if you want.

**Comment by cynicaldevil** [comment]
I added a condition to generate the tuple struct for classes as well.

File: src/ir/comp.rs [hunk 9]
```diff
@@ -481,6 +486,10 @@ impl CompInfo {
         debug!("CompInfo::from_ty({:?}, {:?})", kind, cursor);
 
         let mut ci = CompInfo::new(kind);
+        ci.is_forward_declaration = location.map_or(true, |cur|
```
Comment (emilio) [inline src/ir/comp.rs]: we could write this as:

```rust
ci.is_forward_declaration = location.map_or(true, |cur| {
    match cur.kind() {
        CXCursor_StructDecl |
        CXCursor_UnionDecl |
        CXCursor_ClassDecl => !cur.is_definition(),
        _ => false,
    }
});
```

Which I think it's a bit more legible, what do you think?

File: tests/headers/forward_declared_complex_types.hpp [hunk 10]
```diff
@@ -0,0 +1,15 @@
+struct Foo;
+
+struct Bar {
+    Foo *f;
+};
+
```
Comment (emilio) [inline tests/headers/forward_declared_complex_types.hpp]: Could you also add an empty struct (`struct Baz {}`), and verify that it doesn't produce an opaque enum?

Comment (cynicaldevil) [inline tests/headers/forward_declared_complex_types.hpp]: Added :)

**Comment by emilio** [comment]
@bors-servo r+

Thanks for doing this @cynicaldevil \o/
